app/scripts/lib/rpc-method-middleware/handlers/send-metadata.js

```diff
@@ -0,0 +1,42 @@
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+
+/**
+ * This internal method is used by our external provider to send metadata about
+ * permission subjects so that we can e.g. display a proper name and icon in
+ * our UI.
+ */
+
+const sendMetadata = {
+  methodNames: [MESSAGE_TYPE.SEND_METADATA],
+  implementation: sendMetadataHandler,
+  hookNames: {
+    addSubjectMetadata: true,
+  },
+};
+export default sendMetadata;
+
+/**
+ * @typedef {Record<string, Function>} SendMetadataOptions
+ * @property {Function} addSubjectMetadata - A function that records subject metadata.
+ */
+
+/**
+ * @param {import('json-rpc-engine').JsonRpcRequest<unknown>} req - The JSON-RPC request object.
+ * @param {import('json-rpc-engine').JsonRpcResponse<true>} res - The JSON-RPC response object.
+ * @param {Function} _next - The json-rpc-engine 'next' callback.
+ * @param {Function} end - The json-rpc-engine 'end' callback.
+ * @param {SendMetadataOptions} options
+ */
+function sendMetadataHandler(req, res, _next, end, { addSubjectMetadata }) {
+  const { params } = req;
+  if (params && typeof params === 'object' && !Array.isArray(params)) {
+    const { icon, ...remainingParams } = params;
+
+    if (typeof remainingParams?.name === 'string') {
```

Comment:
This seems like an odd choice :thinking:

Reply:
I agree completely, but that's what the existing implementation is doing: https://github.com/MetaMask/metamask-extension/blob/76a2a9b/app/scripts/controllers/permissions/permissionsMethodMiddleware.js#L74-L82

I suppose the intention is to say that "if there's no name, then this is probably not valid metadata", but there could still be an icon... Maybe we should update the validation, return an error if it fails, and just forward metadata to the controller, name or no?

Reply:
Interesting!

>Maybe we should update the validation, return an error if it fails, and just forward metadata to the controller, name or no?

That sounds better to me

Reply:
Done here: 6804a8fd2